nmmo/task/predicate.py

```diff
@@ -0,0 +1,256 @@
+        # "<class 'nmmo.systems.item.Ration'>", "<class 'nmmo.io.action.Melee'>"
+        tmp_list.append(str(arg)[1:-2].rsplit('.', maxsplit=1)[-1])
+      elif "object at" in str(arg):
+        tmp_list.append(str(arg).split(' object', maxsplit=1)[0].split('.')[-1])
+      elif arg is None:
+        tmp_list.append('Any')
+      else:
+        tmp_list.append(str(arg))
+    return '_'.join(tmp_list).replace(' ', '')
+
+  def __call__(self, gs: GameState) -> bool:
+    """One should describe the code how evaluation is done.
+       LLM might use it to produce goal embedding, which will be
+       used by the RL agent to produce action.
+    """
+    raise NotImplementedError
+
+  def _desc(self, class_type):
+    return {
+      "type": class_type,
+      "name": self.name,
+      "evaluate": self.__call__.__doc__
+    }
+
+  @property
+  def description(self) -> Dict:
+    return self._desc("Predicate")
+
+  def __and__(self, other):
+    return AND(self,other)
+  def __or__(self, other):
+    return OR(self,other)
+  def __invert__(self):
+    return NOT(self)
+  def __rshift__(self,other):
+    return IMPLY(self,other)
+
+
+class AND(Predicate):
+  def __init__(self, *tasks: Predicate):
+    # pylint: disable=super-init-not-called
+    assert len(tasks) > 0
+    self._tasks = tasks
+
+    # the name is AND(task1,task2,task3)
+    self._name = 'AND(' + ','.join([t.name for t in self._tasks]) + ')'
+
+  def __call__(self, gs: GameState) -> bool:
+    """True if all _tasks are evaluated to be True.
+       Otherwise false."""
+    return all(t(gs) for t in self._tasks)
+
+  @property
+  def description(self) -> Dict:
+    desc = self._desc("Conjunction")
+    desc.update({ 'desc_child': ["AND"] + [t.description for t in self._tasks] })
+    return desc
+
+class OR(Predicate):
+  def __init__(self, *tasks: Predicate):
+    # pylint: disable=super-init-not-called
+    assert len(tasks) > 0
+    self._tasks = tasks
+
+    # the name is OR(task1,task2,task3,...)
+    self._name = 'OR(' + ','.join([t.name for t in self._tasks]) + ')'
+
+  def __call__(self, gs: GameState) -> bool:
+    """True if any of _tasks is evaluated to be True.
+       Otherwise false."""
+    return any(t(gs) for t in self._tasks)
+
+  @property
+  def description(self) -> Dict:
+    desc = self._desc("Disjunction")
+    desc.update({ 'desc_child': ["OR"] + [t.description for t in self._tasks] })
+    return desc
+
+class NOT(Predicate):
+  def __init__(self, task: Predicate):
+    # pylint: disable=super-init-not-called
+    self._task = task
+
+    # the name is NOT(task)
+    self._name = f'NOT({self._task.name})'
+
+  def __call__(self, gs: GameState) -> bool:
+    """True if _task is evaluated to be False.
+       Otherwise true."""
+    return not self._task(gs)
+
+  @property
+  def description(self) -> Dict:
+    desc = self._desc("Negation")
+    desc.update({ 'desc_child': ["NOT", self._task.description] })
+    return desc
+
+class IMPLY(Predicate):
```

Comment:
When do we think we'd use this? Since it can be done with NOT and OR, it seems unnessary

Reply:
looked fancy, but removing it.

Reply:
hmm, I'm not sure... Why don't we try with ELM and see what it can make?